File: tests/test_r_guardrails/test_json_provenance.R
# Test: JSON Data Provenance Validation
# ======================================
# Guardrail tests ensuring all JSON exports have proper metadata.
#
# Run: Rscript -e "testthat::test_file('tests/test_r_guardrails/test_json_provenance.R')"
#
# Created: 2026-01-27
# Author: Foundation PLR Team

suppressPackageStartupMessages({
  library(testthat)
  library(jsonlite)
})

# Find project root
find_project_root <- function() {
  markers <- c("pyproject.toml", "CLAUDE.md", ".git")
  dir <- getwd()
  while (dir != dirname(dir)) {
    if (any(file.exists(file.path(dir, markers)))) {
      return(dir)
    }
    dir <- dirname(dir)
  }
  stop("Could not find project root")
}

PROJECT_ROOT <- find_project_root()
R_DATA_DIR <- file.path(PROJECT_ROOT, "outputs/r_data")

# ==============================================================================
# TEST: All JSON exports have metadata
# ==============================================================================

test_that("All JSON exports have metadata section", {
  json_files <- list.files(R_DATA_DIR, pattern = "\\.json$", full.names = TRUE)

  if (length(json_files) == 0) {
    skip("No JSON files found in outputs/r_data/")
  }

  missing_metadata <- c()

  for (json_file in json_files) {
    tryCatch({
      data <- fromJSON(json_file, simplifyVector = FALSE)
      if (is.null(data$metadata)) {
        missing_metadata <- c(missing_metadata, basename(json_file))
      }
    }, error = function(e) {
      warning(sprintf("Could not parse %s: %s", basename(json_file), e$message))
    })
  }

  if (length(missing_metadata) > 0) {
    msg <- sprintf(
      "GUARDRAIL VIOLATION: %d JSON files missing metadata section:\n\n  %s\n\nFIX: Add metadata with generator, data_source, and db_hash.",
      length(missing_metadata),
      paste(missing_metadata, collapse = "\n  ")
    )
    fail(msg)
  }
})

# ==============================================================================
# TEST: All JSON exports have data source hash
# ==============================================================================

test_that("All JSON exports have data_source with db_hash", {
  json_files <- list.files(R_DATA_DIR, pattern = "\\.json$", full.names = TRUE)

  if (length(json_files) == 0) {
    skip("No JSON files found")
  }

  missing_hash <- c()

  for (json_file in json_files) {
    tryCatch({
      data <- fromJSON(json_file, simplifyVector = FALSE)

      # Check for db_hash (may be in different locations)
      has_hash <- FALSE
      if (!is.null(data$metadata$data_source$db_hash)) has_hash <- TRUE
      if (!is.null(data$metadata$db_hash)) has_hash <- TRUE
      if (!is.null(data$metadata$data_source$hash)) has_hash <- TRUE

      if (!has_hash) {
        missing_hash <- c(missing_hash, basename(json_file))
      }
    }, error = function(e) {
      # Skip unparseable files
    })
  }

  if (length(missing_hash) > 0) {
    msg <- sprintf(
      "GUARDRAIL VIOLATION: %d JSON files missing db_hash for provenance tracking:\n\n  %s\n\nFIX: Run python scripts/fix_json_provenance.py or regenerate exports.",
      length(missing_hash),
      paste(missing_hash, collapse = "\n  ")
    )
    fail(msg)
  }
})

# ==============================================================================
# TEST: All JSON exports have generator field
# ==============================================================================

test_that("All JSON exports have generator field", {
  json_files <- list.files(R_DATA_DIR, pattern = "\\.json$", full.names = TRUE)

  if (length(json_files) == 0) {
    skip("No JSON files found")
  }

  missing_generator <- c()

  for (json_file in json_files) {
    tryCatch({
      data <- fromJSON(json_file, simplifyVector = FALSE)
      if (is.null(data$metadata$generator)) {
        missing_generator <- c(missing_generator, basename(json_file))
      }
    }, error = function(e) {
      # Skip unparseable files
    })
  }

  if (length(missing_generator) > 0) {
    msg <- sprintf(
      "GUARDRAIL VIOLATION: %d JSON files missing generator field:\n\n  %s\n\nFIX: Add metadata.generator with the export script path.",
      length(missing_generator),
      paste(missing_generator, collapse = "\n  ")
    )
    fail(msg)
  }
})

# ==============================================================================
# TEST: No synthetic data indicators
# ==============================================================================

test_that("Predictions are from REAL data, not synthetic", {
  json_files <- list.files(R_DATA_DIR, pattern = "prediction.*\\.json$", full.names = TRUE)

  if (length(json_files) == 0) {
    skip("No prediction JSON files found")
  }

  synthetic_indicators <- c("synthetic", "random", "fake", "mock", "test_data")
  suspicious_files <- c()

  for (json_file in json_files) {
    tryCatch({
      content <- readLines(json_file, warn = FALSE)
      content_str <- paste(content, collapse = "\n")

      for (indicator in synthetic_indicators) {
        if (grepl(indicator, content_str, ignore.case = TRUE)) {
          suspicious_files <- c(suspicious_files, basename(json_file))
          break
        }
      }
    }, error = function(e) {
      # Skip unreadable files
    })
  }

  if (length(suspicious_files) > 0) {
    msg <- sprintf(
      "CRITICAL: %d files may contain synthetic data:\n\n  %s\n\nThis is a CRITICAL FAILURE for scientific integrity!",
      length(suspicious_files),
      paste(suspicious_files, collapse = "\n  ")
    )
    fail(msg)
  }
})

# ==============================================================================
# TEST: ROC/RC data has correct schema
# ==============================================================================

test_that("roc_rc_data.json has correct schema", {
  json_path <- file.path(R_DATA_DIR, "roc_rc_data.json")

  if (!file.exists(json_path)) {
    skip("roc_rc_data.json not found - run export_roc_rc_data.py first")
  }

  data <- fromJSON(json_path, simplifyVector = FALSE)

  # Check structure
  expect_true("data" %in% names(data), info = "Missing 'data' section")
  expect_true("configs" %in% names(data$data), info = "Missing 'configs' in data")
  expect_gte(length(data$data$configs), 9, info = "Expected at least 9 configs")

  # Check each config
  for (config in data$data$configs) {
    expect_true(!is.null(config$id), info = sprintf("Config missing 'id'"))
    expect_true(!is.null(config$roc), info = sprintf("Config '%s' missing 'roc'", config$id))
    expect_true(!is.null(config$rc), info = sprintf("Config '%s' missing 'rc'", config$id))

    # ROC curve validity
    if (!is.null(config$roc$fpr)) {
      expect_equal(config$roc$fpr[[1]], 0, tolerance = 0.01,
                   info = "ROC FPR should start at 0")
      expect_equal(tail(config$roc$fpr, 1)[[1]], 1, tolerance = 0.01,
                   info = "ROC FPR should end at 1")
    }
  }
})

# ==============================================================================
# TEST: Selective classification data has correct schema
# ==============================================================================

test_that("selective_classification_data.json has correct schema", {
  json_path <- file.path(R_DATA_DIR, "selective_classification_data.json")

  if (!file.exists(json_path)) {
    skip("selective_classification_data.json not found - run export script first")
  }

  data <- fromJSON(json_path, simplifyVector = FALSE)

  # Check structure
  expect_true("data" %in% names(data))
  expect_true("retention_levels" %in% names(data$data))
  expect_true("configs" %in% names(data$data))

  # Check retention levels
  levels <- unlist(data$data$retention_levels)
  expect_true(all(levels >= 0 & levels <= 1),
              info = "Retention levels must be between 0 and 1")

  # Check configs have required fields
  for (config in data$data$configs) {
    expect_true(!is.null(config$auroc_at_retention))
    expect_true(!is.null(config$net_benefit_at_retention))
    expect_true(!is.null(config$scaled_brier_at_retention))
  }
})

message("\n[test_json_provenance.R] All provenance tests completed.")

File: tests/test_guardrails/test_json_provenance.py
"""
Guardrail Tests: JSON Provenance Metadata

All JSON exports must have complete provenance metadata for reproducibility.
"""

import json
from pathlib import Path

import pytest

pytestmark = [pytest.mark.data, pytest.mark.guardrail]

PROJECT_ROOT = Path(__file__).parent.parent.parent


class TestJSONProvenance:
    """Test that all JSON exports have proper provenance metadata."""

    def get_json_files(self):
        """Get all JSON data files."""
        r_data_dir = PROJECT_ROOT / "data" / "r_data"
        if not r_data_dir.exists():
            return []
        return list(r_data_dir.glob("*.json"))

    def test_all_json_have_metadata(self):
        """All JSON files must have a metadata section."""
        json_files = self.get_json_files()
        if not json_files:
            pytest.skip("No JSON files found in data/r_data/")

        violations = []
        for json_file in json_files:
            try:
                data = json.loads(json_file.read_text())
                if "metadata" not in data:
                    violations.append(
                        {
                            "file": json_file.name,
                            "issue": 'Missing "metadata" key',
                        }
                    )
            except json.JSONDecodeError as e:
                violations.append(
                    {
                        "file": json_file.name,
                        "issue": f"Invalid JSON: {e}",
                    }
                )

        if violations:
            msg = "GUARDRAIL VIOLATION: JSON files missing metadata!\n\n"
            for v in violations:
                msg += f"  {v['file']}: {v['issue']}\n"
            pytest.fail(msg)

    def test_json_have_data_source(self):
        """All JSON files must have data_source in metadata."""
        json_files = self.get_json_files()
        if not json_files:
            pytest.skip("No JSON files found")

        violations = []
        for json_file in json_files:
            try:
                data = json.loads(json_file.read_text())
                metadata = data.get("metadata", {})
                if "data_source" not in metadata:
                    violations.append(
                        {
                            "file": json_file.name,
                            "issue": 'Missing "data_source" in metadata',
                        }
                    )
            except json.JSONDecodeError:
                pass  # Already caught in previous test

        if violations:
            msg = "GUARDRAIL VIOLATION: JSON files missing data_source!\n\n"
            for v in violations:
                msg += f"  {v['file']}: {v['issue']}\n"
            msg += "\nFIX: Add data_source with database path and hash."
            pytest.fail(msg)

    def test_json_have_db_hash(self):
        """All JSON files must have database hash for reproducibility."""
        json_files = self.get_json_files()
        if not json_files:
            pytest.skip("No JSON files found")

        violations = []
        for json_file in json_files:
            try:
                data = json.loads(json_file.read_text())
                data_source = data.get("metadata", {}).get("data_source", {})
                if isinstance(data_source, dict) and "db_hash" not in data_source:
                    # Allow if data_source is a string (simple reference)
                    violations.append(
                        {
                            "file": json_file.name,
                            "issue": 'Missing "db_hash" in data_source',
                        }
                    )
            except json.JSONDecodeError:
                pass

        if violations:
            msg = "GUARDRAIL VIOLATION: JSON files missing db_hash!\n\n"
            for v in violations:
                msg += f"  {v['file']}: {v['issue']}\n"
            msg += "\nFIX: Add hash of source database file for audit trail."
            pytest.fail(msg)

    def test_json_have_generator_info(self):
        """All JSON files should have generator script info."""
        json_files = self.get_json_files()
        if not json_files:
            pytest.skip("No JSON files found")

        violations = []
        for json_file in json_files:
            try:
                data = json.loads(json_file.read_text())
                metadata = data.get("metadata", {})
                if "generator" not in metadata:
                    violations.append(
                        {
                            "file": json_file.name,
                            "issue": 'Missing "generator" in metadata',
                        }
                    )
            except json.JSONDecodeError:
                pass

        if violations:
            msg = "GUARDRAIL VIOLATION: JSON files missing generator info!\n\n"
            for v in violations:
                msg += f"  {v['file']}: {v['issue']}\n"
            msg += "\nFIX: Add generator script path to metadata."
            pytest.fail(msg)

    def test_no_synthetic_data_markers(self):
        """Check that JSON files don't indicate synthetic/fake data."""
        json_files = self.get_json_files()
        if not json_files:
            pytest.skip("No JSON files found")

        violations = []
        suspicious_terms = ["synthetic", "fake", "random", "simulated", "mock"]

        for json_file in json_files:
            content = json_file.read_text().lower()
            for term in suspicious_terms:
                if term in content:
                    violations.append(
                        {
                            "file": json_file.name,
                            "issue": f'Contains suspicious term: "{term}"',
                        }
                    )

        if violations:
            msg = "GUARDRAIL VIOLATION: Possible synthetic data in JSON files!\n\n"
            for v in violations:
                msg += f"  {v['file']}: {v['issue']}\n"
            msg += (
                "\nCRITICAL: All figure data must come from REAL experimental results."
            )
            pytest.fail(msg)
